execution_engine/src/runtime/externals.rs

```diff
@@ -1124,6 +1125,100 @@ where
+                        ApiError::MaxTopicNameSizeExceeded,
+                    )))));
+                }
+
+                let topic_name = self.string_from_mem(topic_name_ptr, topic_name_size)?;
+
+                if operation_size as usize > MessageTopicOperation::max_serialized_len() {
+                    return Err(Trap::from(Error::InvalidMessageTopicOperation));
+                }
+                let topic_operation = self
+                    .t_from_mem(operation_ptr, operation_size)
+                    .map_err(|_e| Trap::from(Error::InvalidMessageTopicOperation))?;
+
+                // only allow managing messages from stored contracts
+                if !self.context.get_entity_key().is_smart_contract_key() {
+                    return Err(Trap::from(Error::InvalidContext));
+                }
+
+                let result = match topic_operation {
+                    MessageTopicOperation::Add => {
+                        self.add_message_topic(topic_name).map_err(Trap::from)?
+                    }
+                };
+
+                Ok(Some(RuntimeValue::I32(api_error::i32_from(result))))
+            }
+            FunctionIndex::EmitMessage => {
+                // args(0) = pointer to the serialized topic name string in wasm memory
+                // args(1) = size of the serialized name string in wasm memory
+                // args(2) = pointer to the serialized message payload in wasm memory
+                // args(3) = size of the serialized message payload in wasm memory
+                let (topic_name_ptr, topic_name_size, message_ptr, message_size) =
+                    Args::parse(args)?;
+
+                // Charge for the call to emit message. This increases for every message emitted
+                // within an execution so we're not using the static value from the wasm config.
+                self.context
+                    .charge_gas(Gas::new(self.context.emit_message_cost()))?;
+                // Charge for parameter weights.
+                self.charge_host_function_call(
+                    &HostFunction::new(0, host_function_costs.emit_message.arguments()),
+                    &[topic_name_ptr, topic_name_size, message_ptr, message_size],
+                )?;
+
+                let limits = self.context.engine_config().wasm_config().messages_limits();
+
+                if topic_name_size > limits.max_topic_name_size() {
+                    return Ok(Some(RuntimeValue::I32(api_error::i32_from(Err(
+                        ApiError::MaxTopicNameSizeExceeded,
+                    )))));
+                }
+
+                if message_size > limits.max_message_size() {
+                    return Ok(Some(RuntimeValue::I32(api_error::i32_from(Err(
+                        ApiError::MessageTooLarge,
+                    )))));
+                }
+
+                let topic_name = self.string_from_mem(topic_name_ptr, topic_name_size)?;
```

Comment:
Same as above

Reply:
Removed serialization in https://github.com/casper-network/casper-node/commit/53483e16590b67de9be9be4d9ba4e6b64d1f22ee although didn't use checked_memory_slice as it's not available yet on the feat-2.0 branch. Will do a follow-up commit after.